src/main/java/graphql/schema/validation/TypeAndFieldRule.java

```diff
@@ -0,0 +1,274 @@
+        if (argumentName.length() >= 2 && argumentName.startsWith("__")) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.InvalidCustomizedNameError,
+                    String.format("Argument name \"%s\" in \"%s-%s\" must not begin with \"__\", which is reserved by GraphQL introspection.", argumentName, typeName, fieldName));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private void assertEnumValueDefinitionName(String typeName, String enumValueDefinitionName, SchemaValidationErrorCollector errorCollector) {
+        if (enumValueDefinitionName.length() >= 2 && enumValueDefinitionName.startsWith("__")) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.InvalidCustomizedNameError,
+                    String.format("EnumValueDefinition \"%s\" in  \"%s\"  must not begin with \"__\", which is reserved by GraphQL introspection.", enumValueDefinitionName, typeName));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private void assertNonNullType(GraphQLType type, SchemaValidationErrorCollector errorCollector) {
+        if (type instanceof GraphQLNonNull && ((GraphQLNonNull) type).getWrappedType() instanceof GraphQLNonNull) {
+            SchemaValidationError schemaValidationError = new SchemaValidationError(SchemaValidationErrorType.NonNullWrapNonNullError,
+                    String.format("Non‐Null type \"%s\" must not wrap another Non‐Null type.", GraphQLTypeUtil.simplePrint(type)));
+            errorCollector.addError(schemaValidationError);
+        }
+    }
+
+    private List<GraphQLNamedType> filterScalarAndIntrospectionTypes(List<GraphQLNamedType> graphQLNamedTypes) {
+        if (graphQLNamedTypes == null || graphQLNamedTypes.isEmpty()) {
+            return Collections.emptyList();
+        }
+
+        return graphQLNamedTypes.stream().filter(type -> !isScalarOrIntrospectionTypes(type)).collect(Collectors.toList());
+    }
+
+    private boolean isScalarOrIntrospectionTypes(GraphQLNamedType type) {
+        if (type instanceof GraphQLScalarType && (
+                type.getName().equals("Int") || type.getName().equals("Float") || type.getName().equals("String") ||
+                        type.getName().equals("Boolean") || type.getName().equals("ID") || type.getName().equals("Long") || type.getName().equals("Short") ||
+                        type.getName().equals("BigInteger") || type.getName().equals("BigDecimal") || type.getName().equals("Char")
+        )) {
+            return true;
+        }
+
+        if (type instanceof GraphQLEnumType && (type.getName().equals("__DirectiveLocation") || type.getName().equals("__TypeKind"))) {
+            return true;
+        }
+
+        if (type instanceof GraphQLObjectType && (
```

Comment:
IntrospectionQuery is not a special type at all (it is just a name).

Reply:
graphql-java framework contains a type named [IntrospectionQuery](https://github.com/graphql-java/graphql-java/blob/10eeacc3979624964ef10ea1078ef84534d9c5cc/src/main/java/graphql/introspection/Introspection.java#L556), which contain field start with '__'.

Reply:
graphql-java framework contains a type named [IntrospectionQuery](https://github.com/graphql-java/graphql-java/blob/10eeacc3979624964ef10ea1078ef84534d9c5cc/src/main/java/graphql/introspection/Introspection.java#L556), which contains fields start with '__'.